### ISSUE TITLE
Package request: PACKAGENAME

### ISSUE THREAD

**Issue by orowith2os** [NixOS/nixpkgs#248000]
Buildstream is used by some projects, specifically freedesktop-sdk, to manage and keep things updated. There's currently no Nix package, so you have to resort to `pip install` or running another distribution's package (in e.g. Distrobox).

> BuildStream is a powerful software integration tool that allows developers to automate the integration of software components including operating systems, and to streamline the software development and production process.

* homepage URL: https://buildstream.build/
* source URL: https://github.com/apache/buildstream
* license: Apache 2.0
* platforms: Linux only, though other systems can use it via container images.

**Comment by nanonyme** [comment]
Note that this implies shipping buildbox stack
* buildbox-common
* buildbox-casd
* buildbox-fuse (technically optional but highly recommended for maximum reproducibility of builds and staging performance reasons)
* buildbox-run-bubblewrap

**Comment by beh-10257** [comment]
since this is going nowhere can any of you help me: https://discourse.nixos.org/t/i-want-to-add-buildstream-from-pip/38810
alsov it would be appreciated if you tag python knowing people

**Comment by nanonyme** [comment]
You can't install it from sources unless you obtain buildbox separately. Also the error is clear, something requires different version of Cython than you have.

**Comment by beh-10257** [comment]
@nanonyme its as you said I guess I have to figure out how to actually create a package nixos to make that merge request
also is it okay to add those packages to nixpkgs
just wanting to know maybe they are not compatible or something

**Comment by nanonyme** [comment]
If you want to study potential deptree, freedesktop-sdk has BuildStream 2 and buildbox build recipes https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/tree/master/elements/components?ref_type=heads

**Comment by shymega** [comment]
As part of my work at Codethink, I have created a initial Buildstream and Buildbox package. I will be upstreaming these to the relevant projects, as well as Nixpkgs. However, for Buildstream, I have only packaged v2. I am not sure if it is worth packaging v1.

These derivations work as expected on `freedesktop-sdk` - with tweaks in the package call.

**Comment by jjardon** [comment]
> However, for Buildstream, I have only packaged v2. I am not sure if it is worth packaging v1.
 
I would not bother with buildstream v1 ; It's deprecated for years and no active project is using AFAIK

**Comment by shymega** [comment]
That's true. I did have a go at packaging, but I've run into an issue where the `buildPythonApplication` function tries to load `libfuse` during the `fixupPhase`, and it's not available, because the wrapped executable with `libfuse` in `$LIBRARY_PATH` isn't around.

**Comment by shymega** [comment]
I've got Buildstream V2 packaged though. The only thing is that I've put a Python dep - `pyroaring` - in the `let` binding. 

@beh-10257 @orowith2os Can you test the Buildstream derivation I've pushed to my fork? Needs a bit of a tidy up, but I'm open to input. The command to run the main `bst` executable is:

> $ nix run github:shymega/nixpkgs/add-buildstream-v2-pkg#buildstream2

Thank you :-)

**Comment by harryfinbow** [comment]
@shymega Just wanted to say that I have been sucessfully using your fork for the last few weeks and was just wondering if you were planning on upstreaming the Buildstream packages at any point in the near future (as it would simplify my flake 😁)?

**Comment by shymega** [comment]
@harryfinbow Yes, I plan to - @jjardon is keen for it, so I need to make a PR first of all for Buildbox, then rebase and make a PR for Buildstream. Or I could squash both in the same PR. Not sure how to go about it.

**Comment by shymega** [comment]
@harryfinbow @jjardon See PR: #383451

**Comment by shymega** [comment]
https://github.com/NixOS/nixpkgs/issues/248000

Partially resolved by https://github.com/NixOS/nixpkgs/pull/383451.

Buildstream & Buildbox are in unstable and 24.11.

Awaiting for the following to be merged (then backported) so that Buildbox derivation is tidier, and Buildstream works with plugins:

- [ ] https://github.com/NixOS/nixpkgs/pull/398620
  - [ ] Merged?
  - [ ] Backported? 
- [ ] https://github.com/NixOS/nixpkgs/pull/398619
  - [ ] Merged?
  - [ ] Backported?

Upon these criterion, I will close this issue as resolved.

**Comment by palfrey** [comment]
Trying to run buildstream at the moment gets you `Error instantiating artifact cache: Did not find "buildbox-casd" in PATH`. I had [a buildstream config that included `pkgs.buildbox` in it's `dependencies`](https://github.com/TraceMachina/nativelink/blob/d319fdae798bc4cfbdce2fcf051b7d1b878644d4/tools/buildstream/bst.nix), not `build-inputs` like https://github.com/nixos/nixpkgs/blob/795f16021c135de737825dceebd65cd44a37a7c6/pkgs/by-name/bu/buildstream/package.nix does and that seemed to resolve the problem.

**Comment by shymega** [comment]
This should fix it.

https://github.com/NixOS/nixpkgs/pull/413743

I'll try and get this merged ASAP. I hadn't noticed the issue as my
Flakes include Buildbox as a separate input.

On 03.06.2025 04:04, Tom Parker-Shemilt wrote:
>palfrey left a comment (NixOS/nixpkgs#248000)
>
>Trying to run buildstream at the moment gets you `Error instantiating artifact cache: Did not find "buildbox-casd" in PATH`. I had [a buildstream config that included `pkgs.buildbox` in it's `dependencies`](https://github.com/TraceMachina/nativelink/blob/d319fdae798bc4cfbdce2fcf051b7d1b878644d4/tools/buildstream/bst.nix), not `build-inputs` like https://github.com/nixos/nixpkgs/blob/795f16021c135de737825dceebd65cd44a37a7c6/pkgs/by-name/bu/buildstream/package.nix does and that seemed to resolve the problem.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/issues/248000#issuecomment-2934704963
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
Best wishes,
--
Dom Rodriguez

**Comment by shymega** [comment]
Still waiting on the above PRs, but we also need https://github.com/NixOS/nixpkgs/pull/419096 to fix the tests.

We're still on 2.4.1 due to an upstream conflict with `protobuf`/grpcio`, but once that's been cleared and a new release made, we can update to 2.5.1.

**Comment by shymega** [comment]
Update: one remaining PR, then a backport once that's merged: https://github.com/NixOS/nixpkgs/pull/421920

This will add Source Provenance to Buildstream.

**Comment by shymega** [comment]
@palfrey Did the latest PRs fix your issue with Buildstream/Buildbox?

**Comment by shymega** [comment]
All pending Buildstream/Buildbox PRs have been merged and backported.

https://github.com/NixOS/nixpkgs/pull/421920 is on its way to 25.05 channels.

The next step is to add Buildstream v2.6.0 once its released, but that's not strictly required for this issue to be marked as complete. Therefore, I will now close this issue as done, and thank everyone for being patient whilst it was being worked on.

Thank you!
